### PR TITLE
Add timestamp support to Pinot connector

### PR DESCRIPTION
**Pull request by xiangfu0**
## Description
Add support for TIMESTAMP data type in Pinot connector.

## Release notes

```markdown
# Pinot Changes
* Add support for `timestamp` type. (#10199)
```

### CODE REVIEW

**Comment by xiangfu0** [comment]
Since https://github.com/trinodb/trino/pull/12145 becomes stale, I just took it over and rebased the latest code.
All credit goes to @ddcprg .

**Comment by ddcprg** [comment]
@xiangfu0 apologies for the delay and many thanks for rebasing the code. Unfortunately I don't have my personal laptop where I am but I'm  happy to either continue the code review on this PR or rebasing your branch into mine if Trino developers prefer to continue reviewing that instead

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public void decode(Supplier<Object> getter, BlockBuilder output)
+    {
+        Object value = getter.get();
+        if (value == null) {
+            output.appendNull();
+        }
+        else {
+            LocalDateTime timestamp;
+            if (value instanceof String) {
```
Comment (Praveen2112) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:41]: How is timestamp internally represented ? As STRING or as LONG ?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java [hunk 2]
```diff
@@ -131,6 +139,37 @@ public static DynamicTable buildFromPql(PinotMetadata pinotMetadata, SchemaTable
         return new DynamicTable(pinotTableName, suffix, selectColumns, filter, groupByColumns, ImmutableList.of(), havingExpression, orderBy, OptionalLong.of(queryContext.getLimit()), getOffset(queryContext), query);
     }
 
+    private static Type toTrinoType(DataSchema.ColumnDataType columnDataType)
```
Comment (Praveen2112) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:142]: Where do we invoke this method ?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java [hunk 3]
```diff
@@ -167,9 +169,20 @@ private static Object convertValue(Type type, Object value)
...
+        }
         return value;
     }
 
+    private static Long toMillis(Long value)
+    {
+        if (value == null) {
+            return null;
+        }
+        return Timestamps.epochMicrosToMillisWithRounding(value);
```
Comment (Praveen2112) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:183]: Can we add a type mapping test like we did for Druid ?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTypeResolver.java [hunk 4]
```diff
@@ -83,4 +93,45 @@ public Type resolveExpressionType(ExpressionContext expression, SchemaTableName
                 throw new PinotException(PINOT_INVALID_PQL_GENERATED, Optional.empty(), format("Unsupported expression: '%s'", expression));
         }
     }
+
+    public static TransformResultMetadata fromTrinoType(Type type)
+    {
+        if (type instanceof ArrayType) {
+            ArrayType arrayType = (ArrayType) type;
```
Comment (Praveen2112) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTypeResolver.java:100]:
```suggestion
        if (type instanceof ArrayType arrayType) {
```

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTypeResolver.java [hunk 5]
```diff
@@ -83,4 +93,45 @@ public Type resolveExpressionType(ExpressionContext expression, SchemaTableName
                 throw new PinotException(PINOT_INVALID_PQL_GENERATED, Optional.empty(), format("Unsupported expression: '%s'", expression));
         }
     }
+
+    public static TransformResultMetadata fromTrinoType(Type type)
```
Comment (Praveen2112) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTypeResolver.java:97]: Where do we invoke this API ?

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java [hunk 6]
```diff
@@ -776,6 +777,7 @@ private static GenericRecord createTestRecord(
                 .set("float_array_col", floatArrayColumn)
                 .set("double_array_col", doubleArrayColumn)
                 .set("long_array_col", longArrayColumn)
+                .set("timestamp_col", updatedAtMillis)
```
Comment (Praveen2112) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java:780]: Can we have a different timestamp here ?

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java [hunk 7]
```diff
@@ -1535,6 +1545,19 @@ public void testLimitPushdown()
...
+    {
+        assertThat(query("SELECT timestamp_col FROM " + ALL_TYPES_TABLE + " WHERE timestamp_col < TIMESTAMP '1971-01-01 00:00:00.000'"))
+                .isFullyPushedDown();
+    }
+
+    @Override
+    public void testCreateTable()
+    {
+        assertQueryFails("CREATE TABLE test_create_table (col INT)", "This connector does not support creating tables");
+    }
```
Comment (Praveen2112) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java:1559]: I think we don't need this test doesn't support  `SUPPORTS_CREATE_TABLE` behavior.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java (hunk 1 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:41]: Pinot broker query gives String and Pinot server query results give numeric value.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java (hunk 2 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java]: removed.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTypeResolver.java (hunk 5 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotTypeResolver.java]: removed

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java (hunk 7 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java:1553]: removed.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java (hunk 6 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java]: added a different value

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java (hunk 3 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:183]: Can you elaborate more?

Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:183]: This value is from presto side, so it's a Long always.

**Comment by xiangfu0** [comment]
Hi @Praveen2112 , thanks for your review. Can u do another pass?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java (hunk 1 above)
Comment (Praveen2112) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:41]: But the decoder is used by `PinotBrokerPageSource` so can we handle only `String` here ?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java (hunk 3 above)
Comment (Praveen2112) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:183]: Actually we have a `TypeMapping` tests for each connector i.e we need something like this - https://github.com/trinodb/trino/blob/master/plugin/trino-druid/src/test/java/io/trino/plugin/druid/TestDruidTypeMapping.java

Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:183]: There is no such typeMapping check here. I suggest to add it later.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java (hunk 1 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:41]: done

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java (hunk 3 above)
Comment (Praveen2112) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:183]: Can we add it as a preparatory commit ? It allows us to test the timestamp of various values.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java (hunk 1 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:41]: Again checked, seems that some functions may return numeric values, so still need to do the type check:
```
Suppressed: java.lang.Exception: SQL: SELECT COUNT(*),  MIN(int_col), MAX(int_col),  MIN(long_col), MAX(long_col), AVG(long_col), SUM(long_col),  MIN(float_col), MAX(float_col), AVG(float_col), SUM(float_col),  MIN(double_col), MAX(double_col), AVG(double_col), SUM(double_col),  MIN(timestamp_col), MAX(timestamp_col)  FROM alltypes
		at io.trino.testing.DistributedQueryRunner.execute(DistributedQueryRunner.java:483)
		... 18 more
```

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java (hunk 3 above)
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:183]: Pinot connector side internally uses this ALL_TYPES table for the testing with kafka.

Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:183]: Added one more test on the time column type for dateTimeField in Pinot.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+                String valueString = (String) value;
```
Comment (martint) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:42]:
```suggestion
            if (value instanceof String stringValue) {
```

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java [hunk 9]
```diff
@@ -167,6 +168,7 @@ private void writeBlock(BlockBuilder blockBuilder, Type columnType, int columnId
             writeBooleanBlock(blockBuilder, columnType, columnIdx);
         }
         else if (javaType.equals(long.class)) {
+            // Applies to timestamp as well since precision is milliseconds
```
[the remark this answers is not in the document]
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java]: The original implementation does the conversion during read block.
I moved the logic to write and just read long value from block.

**Review by ddcprg — APPROVED** [review]
thanks for rebasing and reviving the PR

**Comment by xiangfu0** [comment]
cc: @ebyhr as well

**Comment by xiangfu0** [comment]
Many thanks @martint !

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java (hunk 7 above)
Comment (ebyhr) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java:1562]: We shouldn't override. The base class has the behavior check.
https://github.com/trinodb/trino/blob/02fabc7f3ca515f347c0613e29d44ed14d16bdda/testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorSmokeTest.java#L128-L134

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java [hunk 10]
```diff
@@ -1535,6 +1548,19 @@ public void testLimitPushdown()
                 .isNotFullyPushedDown(LimitNode.class);
     }
 
+    @Test
+    public void testPredicatePushdown()
+    {
+        assertThat(query("SELECT timestamp_col FROM " + ALL_TYPES_TABLE + " WHERE timestamp_col < TIMESTAMP '1971-01-01 00:00:00.000'"))
+                .isFullyPushedDown();
```
Comment (ebyhr) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/BasePinotIntegrationConnectorSmokeTest.java:1555]: The method name is `testPredicatePushdown` but the test is specific to timestamp type. Moving to `testTimestamp` seems better.

**Pull request merged**